File: accelerator-core/src/androidTest/java/com/tokbox/android/accpack/textchat/testbase/BaseTest.java
package com.tokbox.android.accpack.textchat.testbase;


import android.content.Context;
import android.util.Log;
import androidx.test.platform.app.InstrumentationRegistry;
import com.opentok.android.OpentokError;
import com.opentok.android.Session;
import com.opentok.android.Stream;
import com.tokbox.android.accpack.textchat.config.APIConfig;
import com.tokbox.android.otsdkwrapper.wrapper.OTAcceleratorSession;
import org.junit.After;
import org.junit.Assert;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;


public abstract class BaseTest {

    public static String LOGTAG = BaseTest.class.getName();

    protected final static int WAIT_TIME = 75;

    protected Context context;
    protected String sessionId = APIConfig.SESSION_ID;
    protected String apiKey = APIConfig.API_KEY;
    protected String token = APIConfig.TOKEN;

    protected OTAcceleratorSession session;

    protected AtomicBoolean sessionConnected = new AtomicBoolean();
    protected AtomicBoolean sessionError = new AtomicBoolean();

    protected CountDownLatch sessionConnectedLock = new CountDownLatch(1);
    protected CountDownLatch sessionErrorLock = new CountDownLatch(1);

    protected OpentokError sessionLastError = null;

    protected Session.SessionListener sessionListener = new Session.SessionListener() {


        @Override
        public void onStreamReceived(Session session, Stream stream) {
            Log.d(LOGTAG,"Session - onStreamReceived");
        }

        @Override
        public void onError(Session session, OpentokError error) {
            Log.d(LOGTAG,"Session - onError");
        }

        @Override
        public void onStreamDropped(Session session, Stream stream) {
            Log.d(LOGTAG,"Session - onStreamDropped");
        }


        @Override
        public void onDisconnected(Session session) {
            Log.d(LOGTAG,"Session - onDisconnected");
        }

        @Override
        public void onConnected(Session session) {
            Log.d(LOGTAG,"Session - onConnected");
            sessionConnected.set(true);
            sessionConnectedLock.countDown();
        }
    };

    protected Context getContext() {
        return InstrumentationRegistry.getInstrumentation().getContext();
    }

    @After
    public void tearDown() throws Exception {
        sessionConnectedLock = new CountDownLatch(1);
        sessionErrorLock = new CountDownLatch(1);
        sessionConnected.set(false);
        sessionError.set(false);
    }


    protected void waitSessionConnected() throws InterruptedException {
        sessionConnectedLock.await(WAIT_TIME, TimeUnit.SECONDS);
        Assert.assertTrue("session failed to connect", sessionConnected.get());
        Assert.assertFalse(sessionError.get());
        Assert.assertNull(sessionLastError);
    }
}

File: accelerator-core/src/androidTest/java/com/tokbox/accelerator/textchat/BaseTest.kt
package com.tokbox.accelerator.textchat

import android.util.Log
import androidx.test.platform.app.InstrumentationRegistry
import com.opentok.android.OpentokError
import com.opentok.android.Session
import com.opentok.android.Session.SessionListener
import com.opentok.android.Stream
import com.tokbox.accelerator.textchat.config.APIConfig
import com.tokbox.otsdkwrapper.wrapper.OTAcceleratorSession
import org.junit.After
import org.junit.Assert
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

abstract class BaseTest {
    protected var instrumentation = InstrumentationRegistry.getInstrumentation()
    protected var context = instrumentation.context
    protected var apiKey = APIConfig.API_KEY
    protected var sessionId = APIConfig.SESSION_ID
    protected var token = APIConfig.TOKEN
    protected var session: OTAcceleratorSession? = null
    protected var sessionConnected = AtomicBoolean()
    protected var sessionError = AtomicBoolean()
    protected var sessionConnectedLock = CountDownLatch(1)
    protected var sessionErrorLock = CountDownLatch(1)
    protected var sessionLastError: OpentokError? = null
    protected var sessionListener: SessionListener = object : SessionListener {
        override fun onStreamReceived(session: Session, stream: Stream) {
            Log.d(LOGTAG, "Session - onStreamReceived")
        }

        override fun onError(session: Session, error: OpentokError) {
            Log.d(LOGTAG, "Session - onError")
        }

        override fun onStreamDropped(session: Session, stream: Stream) {
            Log.d(LOGTAG, "Session - onStreamDropped")
        }

        override fun onDisconnected(session: Session) {
            Log.d(LOGTAG, "Session - onDisconnected")
        }

        override fun onConnected(session: Session) {
            Log.d(LOGTAG, "Session - onConnected")
            sessionConnected.set(true)
            sessionConnectedLock.countDown()
        }
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        sessionConnectedLock = CountDownLatch(1)
        sessionErrorLock = CountDownLatch(1)
        sessionConnected.set(false)
        sessionError.set(false)
    }

    @Throws(InterruptedException::class)
    protected fun waitSessionConnected() {
        sessionConnectedLock.await(WAIT_TIME.toLong(), TimeUnit.SECONDS)
        Assert.assertTrue("session failed to connect", sessionConnected.get())
        Assert.assertFalse(sessionError.get())
        Assert.assertNull(sessionLastError)
    }

    companion object {
        var LOGTAG = BaseTest::class.java.name
        protected const val WAIT_TIME = 75
    }
}